README.md
# sc-flexy-logger-lib

THE CODE IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND. HMS DOES NOT WARRANT THAT THE FUNCTIONS OF THE CODE WILL MEET YOUR REQUIREMENTS, OR THAT THE OPERATION OF THE CODE WILL BE UNINTERRUPTED OR ERROR-FREE, OR THAT DEFECTS IN IT CAN BE CORRECTED.
---

## [Table of Contents](#table-of-contents)

1. [Description](#description)
2. [Log Level](#log-level)
3. [Socket Logging](#socket-logging)
4. [Developer Documentation](#developer-documentation)
5. [Dependencies](#dependencies)

---

## [Description](#table-of-contents)

Logging library to enable runtime configurable log output in an Ewon Flexy Java application.

## [Log Level](#table-of-contents)

There are seven levels of configurable log level. The logging level is set by the SET_LOG_LEVEL(int) method. Each log level includes the output for the log levels below it (lower numerical value). For example, log level 3 (Warning) includes the output for log level 2 (Serious) and log Level 1 (Critical). All log levels print to the Flexy realtime logs. Negative logging levels write to text files in the /usr directory of the Flexy instead of the realtime log. These text files are named logN.txt where N is an integer.

| LoggingLevel     | Description |
|------------------|---|
| 6, -6 (Trace)    | Exception stack traces |
| 5, -5 (Debug)    | Low level information about the state of the application |
| 4, -4 (Info)     | General Application state information |
| 3, -3 (Warning)  | Issues encountered in the application that are not serious |
| 2, -2 (Serious)  | Serious application log messages (Errors) |
| 1, -1 (Critical) | Critical application log messages (Startup, Initialization, Unrecoverable Error) |
| 0 (None)         | Logging is disabled |

## [Socket Logging](#table-of-contents)
This library supports outputting log information over a raw socket connection on port 766.

Netcat Example: Connect to a Flexy with IP address 10.0.0.53
```console
nc 10.0.0.53 766
```
Note: Socket logging must be enabled using ENABLE_SOCKET_LOG() method.

## [Developer Documentation](#table-of-contents)

Developer documentation is available in Javadoc format found in the release package [https://github.com/hms-networks/sc-flexy-logger-lib/releases](https://github.com/hms-networks/sc-flexy-logger-lib/releases).

## [Dependencies](#table-of-contents)
None

CHANGELOG.md
# Ewon Flexy Logger Library Changelog

## v1.2.1
### Major Changes
- None
### Minor Changes
- Bugfix: Fixed invalid or missing Javadocs

## v1.2
### Major Changes
- Feature: Log queuing of log entries below configured log level
   * Logger.ENABLE_LOG_QUEUE()
   * Logger.DUMP_LOG_QUEUE()
- Feature: Log queuing of SocketLogger logs while disconnected. Messages queued will be outputted upon reconnection.
### Minor Changes
- Bugfix: Fixed infinite recursive loop when socket connection lost.
- Updated README.md for SocketLogger

## v1.1
### Major Changes
- Feature: Added socket logging functionality
### Minor Changes
- Feature: Added methods to enable and disable logging to Ewon Flexy realtime logs
   * Logger.ENABLE_REALTIME_LOG()
   * Logger.DISABLE_REALTIME_LOG()

## v1.0
Initial release

src/com/hms_networks/americas/sc/logging/LogQueue.java
package com.hms_networks.americas.sc.logging;

/**
 * LogQueue.java
 *
 * <p>Provides an queue mechanism for log entries which minimizes dynamic memory allocation.
 *
 * @since 1.2
 * @author HMS Networks, MU Americas Solution Center
 */
public class LogQueue {

  /** Constant for max string size */
  private static final int DEFAULT_STRING_MAX_SIZE_CHARS = 255;

  /** 2D char array to represent a queue of char arrays */
  private char[][] logQueue;

  /** Array containing the size of each log entry in the queue */
  private int[] logQueueLengths;

  /** Number of entries the queue can hold */
  private int queueDepth;

  /** Current tail index of queue */
  private int currTailIndex;

  /** Current number of entries in queue */
  private int currQueueSize;

  /**
   * Creates a new LogQueue
   *
   * @param queueSize number of elements the queue can hold
   */
  public LogQueue(int queueSize) {
    currQueueSize = 0;
    currTailIndex = -1;
    queueDepth = queueSize;
    logQueue = new char[queueDepth][DEFAULT_STRING_MAX_SIZE_CHARS];
    logQueueLengths = new int[queueDepth];
  }

  /**
   * Queues a new string
   *
   * @param logEntry the string to be enqueued
   */
  public void addLogEntry(String logEntry) {

    // Increment queue index
    currTailIndex++;

    // Rotate around to beginning of array if the end is reached
    if (currTailIndex >= queueDepth) {
      currTailIndex = 0;
    }

    // Increment queue size if not full
    if (currQueueSize < queueDepth) {
      currQueueSize++;
    }

    // Store the length of string
    logQueueLengths[currTailIndex] = logEntry.length();

    // Limit log entry to max size
    if (logQueueLengths[currTailIndex] > DEFAULT_STRING_MAX_SIZE_CHARS) {
      logQueueLengths[currTailIndex] = DEFAULT_STRING_MAX_SIZE_CHARS;
      logEntry = logEntry.substring(0, DEFAULT_STRING_MAX_SIZE_CHARS);
    }

    logQueue[currTailIndex] = logEntry.toCharArray();
  }

  /**
   * Method to get head string of queue
   *
   * @return the head string of the queue
   */
  public String poll() {

    String res = null;
    if (currQueueSize > 0) {
      // Compute the head index
      int headIndex = (currTailIndex + queueDepth - currQueueSize + 1) % queueDepth;

      res = String.valueOf(logQueue[headIndex]).substring(0, logQueueLengths[headIndex]);
      currQueueSize--;
    }
    return res;
  }

  /**
   * Method to check if the queue is empty
   *
   * @return true if queue is empty
   */
  public boolean isEmpty() {
    boolean res = false;
    if (currQueueSize == 0) {
      res = true;
    }
    return res;
  }

  /**
   * Method to check if the queue is fully loaded
   *
   * @return true if queue is full
   */
  public boolean isFull() {
    boolean res = false;
    if (currQueueSize == queueDepth) {
      res = true;
    }
    return res;
  }
}


src/com/hms_networks/americas/sc/logging/SocketLogger.java
package com.hms_networks.americas.sc.logging;

import java.io.DataOutputStream;
import java.io.IOException;
import javax.microedition.io.ConnectionNotFoundException;
import javax.microedition.io.Connector;
import javax.microedition.io.ServerSocketConnection;
import javax.microedition.io.SocketConnection;

/**
 * SocketLogger.java
 *
 * <p>Provides an interface for logging via socket connection.
 *
 * @since 1.1
 * @author HMS Networks, MU Americas Solution Center
 */
public class SocketLogger implements Runnable {

  /** Constant for socket port */
  private static final String SOCKET_PORT = "766";

  /** Constant for socket fatal error state */
  private static final int STATE_FATAL_ERROR = -1;

  /** Constant for socket disconnected state */
  private static final int STATE_DISCONNECTED = 0;

  /** Constant for socket connecting state */
  private static final int STATE_CONNECTING = 1;

  /** Constant for socket connected state */
  private static final int STATE_CONNECTED = 2;

  /** Constant for socket delay option */
  private static final byte SOCKET_DELAY_SEC = 0;

  /** Constant for socket linger option */
  private static final byte SOCKET_LINGER_SEC = 5;

  /** Constant for socket keep alive option */
  private static final byte SOCKET_KEEPALIVE_SEC = 15;

  /** Constant for socket receive buffer size option */
  private static final byte SOCKET_RCVBUF_BYTES = 1;

  /** Constant for socket send buffer size option */
  private static final byte SOCKET_SNDBUF_BYTES = 127;

  /** Server socket connection used for logging */
  private static ServerSocketConnection socketServerConnection;

  /** Underlying socket connection used for logging */
  private static SocketConnection socketConnection;

  /** Data output stream for adding log entries to */
  private static DataOutputStream dataOutputStream;

  /** Socket connection state variable */
  private static int socketConnectionState = STATE_DISCONNECTED;

  /** Socket log queue for storing logs while not connected */
  private static LogQueue socketLogQueue;

  /** Number of log entries able to be queued */
  private static final int SOCKET_LOG_QUEUE_DEPTH = 50;

  /** Open the socket server connection */
  private static void OPEN() {
    if (socketLogQueue == null) {
      socketLogQueue = new LogQueue(SOCKET_LOG_QUEUE_DEPTH);
    }

    if (socketConnectionState == STATE_DISCONNECTED) {
      socketConnectionState = STATE_CONNECTING;
      Logger.LOG_DEBUG("SocketLogger connection starting.");
      // Create the server listening socket
      try {
        socketServerConnection =
            (ServerSocketConnection) Connector.open("socket://:" + SOCKET_PORT);
      } catch (ConnectionNotFoundException e) {
        Logger.LOG_DEBUG("SocketLogger connection could not open (ConnectionNotFoundException).");
        socketConnectionState = STATE_DISCONNECTED;
      } catch (IllegalArgumentException e) {
        Logger.LOG_DEBUG("SocketLogger connection could not open (IllegalArgumentException).");
        socketConnectionState = STATE_DISCONNECTED;
      } catch (SecurityException e) {
        Logger.LOG_DEBUG("SocketLogger connection could not open (SecurityException).");
        socketConnectionState = STATE_DISCONNECTED;
      } catch (IOException e) {
        Logger.LOG_DEBUG("SocketLogger connection could not open (IOException).");
        socketConnectionState = STATE_DISCONNECTED;
      }
      if (socketConnectionState == STATE_CONNECTING) {
        SocketLogger sl = new SocketLogger();
        Thread connectThread = new Thread(sl);
        connectThread.start();
      }
    }
  }

  /** Close the socket server connection */
  private static void CLOSE() {
    socketConnectionState = STATE_DISCONNECTED;
    Logger.LOG_DEBUG("SocketLogger connection closing.");

    // Try to close all elements of connection.
    try {
      dataOutputStream.close();
    } catch (IOException e) {
      Logger.LOG_DEBUG("SocketLogger data output stream close failed.");
    }
    try {
      socketConnection.close();
    } catch (IOException e) {
      Logger.LOG_DEBUG("SocketLogger socket connection close failed.");
    }
    try {
      socketServerConnection.close();
    } catch (IOException e) {
      Logger.LOG_DEBUG("SocketLogger socket server connection close failed.");
    }
  }

  /**
   * Run method
   *
   * <p>Handles waiting for a incoming connection and setting up the socket connection
   */
  public void run() {
    try {
      // Wait for a incoming connection
      socketConnection = (SocketConnection) socketServerConnection.acceptAndOpen();

      // Set socket connection options
      socketConnection.setSocketOption(SocketConnection.DELAY, SOCKET_DELAY_SEC);
      socketConnection.setSocketOption(SocketConnection.LINGER, SOCKET_LINGER_SEC);
      socketConnection.setSocketOption(SocketConnection.KEEPALIVE, SOCKET_KEEPALIVE_SEC);
      socketConnection.setSocketOption(SocketConnection.RCVBUF, SOCKET_RCVBUF_BYTES);
      socketConnection.setSocketOption(SocketConnection.SNDBUF, SOCKET_SNDBUF_BYTES);

      // Open a data output stream to add log information to
      dataOutputStream = socketConnection.openDataOutputStream();
    } catch (IllegalArgumentException e) {
      // Socket options are incorrect, set state to fatal error.
      Logger.LOG_DEBUG("SocketLogger invalid socket options specified.");
      socketConnectionState = STATE_FATAL_ERROR;
    } catch (IOException e) {
      Logger.LOG_DEBUG("SocketLogger connection could not start.");
      socketConnectionState = STATE_DISCONNECTED;
    }

    if (socketConnectionState == STATE_CONNECTING) {
      // Connection has been established
      socketConnectionState = STATE_CONNECTED;
      outputQueue();
    }
  }

  /** Outputs all queued log entries if possible */
  private void outputQueue() {
    while (!socketLogQueue.isEmpty() && socketConnectionState == STATE_CONNECTED) {
      String logEntry = socketLogQueue.poll();
      try {
        writeStringToSocket(logEntry);
      } catch (IOException e) {

        // Put log entry back in queue
        socketLogQueue.addLogEntry(logEntry);

        CLOSE();
        OPEN();
      }
    }
  }

  /**
   * Log a string to the socket connection if connected. If no connection exists this method will
   * start one.
   *
   * @param logEntry string to log
   */
  public static void LOG(String logEntry) {
    boolean wasLogged = false;
    if (logEntry.length() > 0) {
      // Format log entries "Timestamp: logEntry\n"
      String logEntryFormatted = System.currentTimeMillis() + ": " + logEntry + "\n";

      if (socketConnectionState == STATE_CONNECTED) {
        try {
          writeStringToSocket(logEntryFormatted);
          wasLogged = true;
        } catch (IOException e) {
          CLOSE();
          OPEN();
        }
      } else if (socketConnectionState == STATE_DISCONNECTED) {
        // State is disconnected, open socket connection
        OPEN();
      }
      if (!wasLogged) {
        // Log entry not sent, put log entry in queue
        socketLogQueue.addLogEntry(logEntryFormatted);
      }
    }
  }

  /**
   * Outputs a string on established socket connection
   *
   * @param s string to output
   * @throws IOException if the socket connection is broken
   * @since 1.2
   */
  private static void writeStringToSocket(String s) throws IOException {
    // Convert string to byte array
    byte[] outputByteArray = s.getBytes("UTF-8");
    dataOutputStream.write(outputByteArray);
  }
}


src/com/hms_networks/americas/sc/logging/Logger.java
package com.hms_networks.americas.sc.logging;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;

/**
 * Logger.java
 *
 * <p>Provides an interface for managing log output using log levels.
 *
 * @since 1.0
 * @author HMS Networks, MU Americas Solution Center
 */
public class Logger {

  /** Constant for trace log level */
  public static final int LOG_LEVEL_TRACE = 6;

  /** Constant for debug log level */
  public static final int LOG_LEVEL_DEBUG = 5;

  /** Constant for info log level */
  public static final int LOG_LEVEL_INFO = 4;

  /** Constant for warning log level */
  public static final int LOG_LEVEL_WARN = 3;

  /** Constant for serious log level */
  public static final int LOG_LEVEL_SERIOUS = 2;

  /** Constant for critical log level */
  public static final int LOG_LEVEL_CRITICAL = 1;

  /** Constant for no log level */
  public static final int LOG_LEVEL_NONE = 0;

  /** Constant for the max size (in chars) of log file */
  private static final int MAX_CHARS_PER_FILE = 25000;

  /** Constant for the max number of log files */
  private static final int MAX_NUM_LOG_FILES = 10;

  /** Constant for the max number of queued logs */
  private static final int MAX_NUM_QUEUED_LOGS = 25;

  /** Constant for the log file path */
  private static final String LOG_FILE_PATH = "/usr/";

  /** Constant for the log file name */
  private static final String LOG_FILE_NAME = "log";

  /** Constant for the log file extension */
  private static final String LOG_FILE_EXTENSION = ".txt";

  /** Stored logging level. Default is DEBUG */
  private static int loggingLevel = LOG_LEVEL_DEBUG;

  /** Current log file index */
  private static int currFileNumber = 1;

  /** Number of chars written to current log file */
  private static long numCharsWrittenToFile = 0;

  /** Indicator if logging to realtime logs */
  private static boolean isLoggingToRealtime = true;

  /** Indicator if logging to realtime logs */
  private static boolean isLoggingToSocket = false;

  /** Indicator if logging to file */
  private static boolean isLoggingToFile = false;

  /** Indicator if queuing logs */
  private static boolean isQueueing = false;

  /** Queue of unprinted logs */
  private static LogQueue logQueue;

  /**
   * Enable logging to Flexy's realtime logs
   *
   * @since 1.1
   */
  public static void ENABLE_REALTIME_LOG() {
    isLoggingToRealtime = true;
  }

  /**
   * Disable logging to Flexy's realtime logs
   *
   * @since 1.1
   */
  public static void DISABLE_REALTIME_LOG() {
    isLoggingToRealtime = false;
  }

  /**
   * Enable logging to socket connection
   *
   * @since 1.1
   */
  public static void ENABLE_SOCKET_LOG() {
    isLoggingToSocket = true;
  }

  /**
   * Disable logging to socket connection
   *
   * @since 1.1
   */
  public static void DISABLE_SOCKET_LOG() {
    isLoggingToSocket = false;
  }

  /**
   * Enable queuing unprinted log messages
   *
   * @since 1.2
   */
  public static void ENABLE_LOG_QUEUE() {
    isQueueing = true;
    logQueue = new LogQueue(MAX_NUM_QUEUED_LOGS);
  }

  /**
   * Set the log level for this class
   *
   * @param level desired log level
   * @return true if successful
   */
  public static boolean SET_LOG_LEVEL(int level) {
    boolean retval = false;

    if (level < 0) {
      isLoggingToFile = true;
      level = Math.abs(level);
    }

    if (level >= LOG_LEVEL_NONE && level <= LOG_LEVEL_TRACE) {
      loggingLevel = level;
      retval = true;

      if (isLoggingToFile) {
        /* Look for an unused log file index and set the currFileNumber to that index. */
        for (int i = 1; i <= MAX_NUM_LOG_FILES; i++) {
          try {
            File file = new File(LOG_FILE_PATH + LOG_FILE_NAME + i + LOG_FILE_EXTENSION);
            if (!file.exists()) {
              currFileNumber = i;
              break;
            }
          } catch (Exception e) {
            /*
             * Issue with access to log file so use realtime log to print error rather than
             * use Logger functionality.
             */
            System.out.println("Error: Could not access stored log files!");
            e.printStackTrace();
          }
        }

        /* Delete the next file in the log entries. This marks where the application left off. */
        int nextFileNumber = currFileNumber + 1;

        if (nextFileNumber > MAX_NUM_LOG_FILES) {
          nextFileNumber = 1;
        }

        try {
          File nextLogFile =
              new File(LOG_FILE_PATH + LOG_FILE_NAME + nextFileNumber + LOG_FILE_EXTENSION);
          nextLogFile.delete();
        } catch (Exception e) {
          Logger.LOG_SERIOUS("Could not delete next log file in sequence!");
          Logger.LOG_EXCEPTION(e);
        }
      }

      Logger.LOG_INFO("Set logging level to " + level);
    }
    return retval;
  }

  /**
   * Log an exception (uses TRACE log level)
   *
   * @param e exception to log
   */
  public static void LOG_EXCEPTION(Exception e) {
    if (loggingLevel == LOG_LEVEL_TRACE) {
      StringWriter exceptionStringWriter = new StringWriter();
      e.printStackTrace(new PrintWriter(exceptionStringWriter));
      String exceptionAsString = exceptionStringWriter.toString();
      LOG(loggingLevel, exceptionAsString);
    }
  }

  /**
   * Log a string using the specified log level
   *
   * @param level desired log level
   * @param logString string to log
   */
  public static void LOG(int level, String logString) {
    if (level <= loggingLevel) {
      if (isLoggingToRealtime) {
        System.out.println(logString);
      }
      if (isLoggingToFile) {
        LOG_TO_FILE(logString);
      }
      if (isLoggingToSocket) {
        SocketLogger.LOG(logString);
      }
    } else if (isQueueing) {
      logQueue.addLogEntry(logString);
    }
  }

  public static void LOG_TO_FILE(String logString) {

    boolean isAppendEnabled = true;

    /*
     * Check if the next file should be opened.
     * If so delete the next file in the sequence holding the most out of
     * date log entries. This missing file will be used on startup to be
     * the log file written first. This maintains a linear log history of
     * the most current events.
     */
    if ((numCharsWrittenToFile + logString.length()) > MAX_CHARS_PER_FILE) {
      int nextFileNumber = currFileNumber + 2;
      currFileNumber++;

      if (nextFileNumber > MAX_NUM_LOG_FILES) {
        nextFileNumber = 1;
      }

      /* Delete the next file in the log entries. This marks where the application left off. */
      try {
        File nextLogFile =
            new File(LOG_FILE_PATH + LOG_FILE_NAME + nextFileNumber + LOG_FILE_EXTENSION);
        nextLogFile.delete();
      } catch (Exception e) {
        Logger.LOG_SERIOUS("Could not delete next log file in sequence!");
        Logger.LOG_EXCEPTION(e);
      }

      if (currFileNumber > MAX_NUM_LOG_FILES) {
        currFileNumber = 1;
      }

      /* Reset num char written counter for new file */
      numCharsWrittenToFile = 0;

      /* This will be the first line in the file so there is no need to append */
      isAppendEnabled = false;
    }

    FileWriter logFileWriter = null;
    BufferedWriter logBufferedWriter = null;

    /* Build a log entry string (timestamp + logString) */
    String formattedLogString = ("[" + System.currentTimeMillis() + "] " + logString + "\n");

    /* Write the log entry to the file */
    try {

      /* Open the log file */
      File logFile = new File(LOG_FILE_PATH + LOG_FILE_NAME + currFileNumber + LOG_FILE_EXTENSION);

      logFileWriter = new FileWriter(logFile, isAppendEnabled);
      logBufferedWriter = new BufferedWriter(logFileWriter);

      logBufferedWriter.write(formattedLogString);

      logBufferedWriter.close();
      logFileWriter.close();

      numCharsWrittenToFile += formattedLogString.length();
    } catch (IOException e) {
      /*
       * Do not attempt to call LOG_EXCEPTION() here as it could cause an infinite loop
       * due to recursion. Write stacktrace to realtime log.
       */
      e.printStackTrace();
    }
  }

  /**
   * Log a string with the debug log level
   *
   * @param logString string to log
   */
  public static void LOG_DEBUG(String logString) {
    LOG(LOG_LEVEL_DEBUG, logString);
  }

  /**
   * Log a string with the info log level
   *
   * @param logString string to log
   */
  public static void LOG_INFO(String logString) {
    LOG(LOG_LEVEL_INFO, logString);
  }

  /**
   * Log a string with the warning log level
   *
   * @param logString string to log
   */
  public static void LOG_WARN(String logString) {
    LOG(LOG_LEVEL_WARN, logString);
  }

  /**
   * Log a string with the serious log level
   *
   * @param logString string to log
   */
  public static void LOG_SERIOUS(String logString) {
    LOG(LOG_LEVEL_SERIOUS, logString);
  }

  /**
   * Log a string with the critical log level
   *
   * @param logString string to log
   */
  public static void LOG_CRITICAL(String logString) {
    LOG(LOG_LEVEL_CRITICAL, logString);
  }

  /**
   * Output all queued logs at critical level
   *
   * @param logString string to log
   */
  public static void DUMP_LOG_QUEUE() {
    while (!logQueue.isEmpty()) {
      LOG(LOG_LEVEL_CRITICAL, logQueue.poll());
    }
  }
}
